src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -496,182 +1255,310 @@ bool emberAfThermostatClusterSetActivePresetRequestCallback(
-        MaxHeatSetpointLimit = AbsMaxHeatSetpointLimit;
+            // If the presetHandle for the pending preset is not null and a matching preset is not found in the
+            // presets attribute list, return NOT_FOUND.
+            if (!foundMatchingPresetInPresets)
+            {
+                return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::NotFound);
+            }
 
-    if (HeatingSetpoint < MinHeatSetpointLimit)
-        HeatingSetpoint = MinHeatSetpointLimit;
+            // Find the number of presets in the pending preset list that match the preset handle. If there are duplicate
+            // entries, return CONSTRAINT_ERROR.
+            uint8_t count = CountPresetsInPendingListWithPresetHandle(delegate, pendingPreset.GetPresetHandle().Value());
+            if (count > 1)
+            {
+                return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::ConstraintError);
+            }
+        }
 
-    if (HeatingSetpoint > MaxHeatSetpointLimit)
-        HeatingSetpoint = MaxHeatSetpointLimit;
+        // If the preset is found in the  presets attribute list and the preset is builtIn in the pending presets list
+        //  but not in the presets attribute list, return UNSUPPORTED_ACCESS.
+        if (foundMatchingPresetInPresets && (IsBuiltIn(pendingPreset) && !IsBuiltIn(matchingPreset)))
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::UnsupportedAccess);
+        }
 
-    return HeatingSetpoint;
-}
+        // If the preset is found in the  presets attribute list and the preset is builtIn in the presets attribute
+        //  but not in the pending presets list, return UNSUPPORTED_ACCESS.
+        if (foundMatchingPresetInPresets && (!IsBuiltIn(pendingPreset) && IsBuiltIn(matchingPreset)))
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::UnsupportedAccess);
+        }
 
-int16_t EnforceCoolingSetpointLimits(int16_t CoolingSetpoint, EndpointId endpoint)
-{
-    // Optional Mfg supplied limits
-    int16_t AbsMinCoolSetpointLimit = kDefaultAbsMinCoolSetpointLimit;
-    int16_t AbsMaxCoolSetpointLimit = kDefaultAbsMaxCoolSetpointLimit;
+        // If the presetScenario is not found in the preset types, return CONSTRAINT_ERROR.
+        PresetScenarioEnum presetScenario = pendingPreset.GetPresetScenario();
+        if (!PresetScenarioExistsInPresetTypes(delegate, presetScenario))
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::ConstraintError);
+        }
 
-    // Optional User supplied limits
-    int16_t MinCoolSetpointLimit = kDefaultMinCoolSetpointLimit;
-    int16_t MaxCoolSetpointLimit = kDefaultMaxCoolSetpointLimit;
+        // If the preset type for the preset scenario does not support name and a name is specified, return CONSTRAINT_ERROR.
```

Comment:
```suggestion
        // If the preset type for the preset scenario does not support name and a names is specified, return CONSTRAINT_ERROR.
```
No, I just had this wrong last time; deleted the wrong "s".  "does not support names" is right.